{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "from statistics import mean\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_with_functions = pd.read_csv(\"./Data/joined_tables.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_cleaned = proteins_with_functions.drop_duplicates(subset=\"structureId\",keep=\"first\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence motif identification\n",
    "\n",
    "I assumed that a good way to extract extra information from the data is to identify motifs that may be present in the sequences.\n",
    "\n",
    "First, the first few amino acids at the beginning and end of each sequence are analysed. It is a simplified version of motif discovery which could have been executed on the whole sequence. This simple approach is reasonable, since in the polypeptide chains some signal peptides are often present at the N or C end of a protein. It made me think that to some extent it could be possible to distinguish protein types based on the first few amino acids in the sequence, as well as some short motifs at the end. \n",
    "\n",
    "The simplest way to check that was to slice first the first few amino acids, deduplicating sub-sequences, encoding them, and using them as a multi-class variable. To avoid increasing dimensionality, frequency encoding was initially attempted, however, this method produced many classes with a count = 1 and as a result it was not suitable for such case, so numerical encoding was used instead, which was eventually transformed into a one-hot encoding."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structureId</th>\n",
       "      <th>chainId</th>\n",
       "      <th>sequence</th>\n",
       "      <th>residueCount</th>\n",
       "      <th>len</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>...</th>\n",
       "      <th>T</th>\n",
       "      <th>U</th>\n",
       "      <th>V</th>\n",
       "      <th>W</th>\n",
       "      <th>Y</th>\n",
       "      <th>X</th>\n",
       "      <th>B</th>\n",
       "      <th>Z</th>\n",
       "      <th>classification</th>\n",
       "      <th>chainCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4LZV</td>\n",
       "      <td>A</td>\n",
       "      <td>LIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEEL...</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>0.030864</td>\n",
       "      <td>0.061728</td>\n",
       "      <td>0.098765</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>...</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.012346</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSPORT PROTEIN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4FK5</td>\n",
       "      <td>A</td>\n",
       "      <td>GAAAAMSICPHIQQVFQNEKSKDGVLKTCNAARYILNHSVPKEKFL...</td>\n",
       "      <td>767</td>\n",
       "      <td>476</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.054622</td>\n",
       "      <td>0.048319</td>\n",
       "      <td>0.048319</td>\n",
       "      <td>0.052521</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.010504</td>\n",
       "      <td>0.033613</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HYDROLASE</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4DOY</td>\n",
       "      <td>A</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTLSPEKQHVRPRDAADNDPVAVARG...</td>\n",
       "      <td>3496</td>\n",
       "      <td>437</td>\n",
       "      <td>0.125858</td>\n",
       "      <td>0.002288</td>\n",
       "      <td>0.059497</td>\n",
       "      <td>0.054920</td>\n",
       "      <td>0.032037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.061785</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.066362</td>\n",
       "      <td>0.022883</td>\n",
       "      <td>0.025172</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2XZK</td>\n",
       "      <td>A</td>\n",
       "      <td>INDPAKSAAPYHDEFPLFRSANMASPDKLSTGIGFHSFRIPAVVRT...</td>\n",
       "      <td>772</td>\n",
       "      <td>386</td>\n",
       "      <td>0.085492</td>\n",
       "      <td>0.002591</td>\n",
       "      <td>0.072539</td>\n",
       "      <td>0.033679</td>\n",
       "      <td>0.036269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.077720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.049223</td>\n",
       "      <td>0.025907</td>\n",
       "      <td>0.033679</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HYDROLASE</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4WR9</td>\n",
       "      <td>A</td>\n",
       "      <td>AYLDEELQTELYEIKHQILQTMGVLSLQGSMLSVGDKVFSTNGQSV...</td>\n",
       "      <td>148</td>\n",
       "      <td>148</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.087838</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.060811</td>\n",
       "      <td>0.013514</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BINDING PROTEIN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117508</th>\n",
       "      <td>3QAH</td>\n",
       "      <td>A</td>\n",
       "      <td>MGSHHHHHHHHGSDYDIPTTENLYFQGSTKVKYVDKIHIGNYEIDA...</td>\n",
       "      <td>304</td>\n",
       "      <td>304</td>\n",
       "      <td>0.023026</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.039474</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042763</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.023026</td>\n",
       "      <td>0.075658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSFERASE</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117509</th>\n",
       "      <td>4M0P</td>\n",
       "      <td>A</td>\n",
       "      <td>MPNIKIFSGSSHQDLSQKIADRLGLELGKVVTKKFSNQETCVEIGE...</td>\n",
       "      <td>652</td>\n",
       "      <td>326</td>\n",
       "      <td>0.095092</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>0.067485</td>\n",
       "      <td>0.052147</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.085890</td>\n",
       "      <td>0.006135</td>\n",
       "      <td>0.015337</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSFERASE</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117510</th>\n",
       "      <td>4NPM</td>\n",
       "      <td>A</td>\n",
       "      <td>DESEYEERRDAEARRVKSGIKQASIFTLEECARIEAKIDEVVAKAD...</td>\n",
       "      <td>500</td>\n",
       "      <td>250</td>\n",
       "      <td>0.072000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.068000</td>\n",
       "      <td>0.076000</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088000</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.032000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117511</th>\n",
       "      <td>2AF0</td>\n",
       "      <td>A</td>\n",
       "      <td>ADLGTENLYFQSMKPSPEEAQLWSEAFDELLASKYGLAAFRAFLKS...</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>0.089041</td>\n",
       "      <td>0.027397</td>\n",
       "      <td>0.041096</td>\n",
       "      <td>0.116438</td>\n",
       "      <td>0.082192</td>\n",
       "      <td>...</td>\n",
       "      <td>0.054795</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006849</td>\n",
       "      <td>0.013699</td>\n",
       "      <td>0.041096</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SIGNALING PROTEIN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117512</th>\n",
       "      <td>2XDP</td>\n",
       "      <td>A</td>\n",
       "      <td>SMCEKVISVGQTVITKHRNTRYYSCRVMAVTSQTFYEVMFDDGSFS...</td>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.073171</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.097561</td>\n",
       "      <td>0.008130</td>\n",
       "      <td>0.056911</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117513 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       structureId chainId                                           sequence  \\\n",
       "0             4LZV       A  LIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEEL...   \n",
       "1             4FK5       A  GAAAAMSICPHIQQVFQNEKSKDGVLKTCNAARYILNHSVPKEKFL...   \n",
       "2             4DOY       A  MGSSHHHHHHSSGLVPRGSHMTLSPEKQHVRPRDAADNDPVAVARG...   \n",
       "3             2XZK       A  INDPAKSAAPYHDEFPLFRSANMASPDKLSTGIGFHSFRIPAVVRT...   \n",
       "4             4WR9       A  AYLDEELQTELYEIKHQILQTMGVLSLQGSMLSVGDKVFSTNGQSV...   \n",
       "...            ...     ...                                                ...   \n",
       "117508        3QAH       A  MGSHHHHHHHHGSDYDIPTTENLYFQGSTKVKYVDKIHIGNYEIDA...   \n",
       "117509        4M0P       A  MPNIKIFSGSSHQDLSQKIADRLGLELGKVVTKKFSNQETCVEIGE...   \n",
       "117510        4NPM       A  DESEYEERRDAEARRVKSGIKQASIFTLEECARIEAKIDEVVAKAD...   \n",
       "117511        2AF0       A  ADLGTENLYFQSMKPSPEEAQLWSEAFDELLASKYGLAAFRAFLKS...   \n",
       "117512        2XDP       A  SMCEKVISVGQTVITKHRNTRYYSCRVMAVTSQTFYEVMFDDGSFS...   \n",
       "\n",
       "        residueCount  len         A         C         D         E         F  \\\n",
       "0                162  162  0.092593  0.030864  0.061728  0.098765  0.024691   \n",
       "1                767  476  0.042017  0.054622  0.048319  0.048319  0.052521   \n",
       "2               3496  437  0.125858  0.002288  0.059497  0.054920  0.032037   \n",
       "3                772  386  0.085492  0.002591  0.072539  0.033679  0.036269   \n",
       "4                148  148  0.054054  0.027027  0.067568  0.087838  0.027027   \n",
       "...              ...  ...       ...       ...       ...       ...       ...   \n",
       "117508           304  304  0.023026  0.026316  0.052632  0.055921  0.039474   \n",
       "117509           652  326  0.095092  0.027607  0.067485  0.052147  0.027607   \n",
       "117510           500  250  0.072000  0.020000  0.068000  0.076000  0.040000   \n",
       "117511           146  146  0.089041  0.027397  0.041096  0.116438  0.082192   \n",
       "117512           123  123  0.040650  0.024390  0.073171  0.081301  0.048780   \n",
       "\n",
       "        ...         T    U         V         W         Y    X    B    Z  \\\n",
       "0       ...  0.049383  0.0  0.055556  0.012346  0.024691  0.0  0.0  0.0   \n",
       "1       ...  0.042017  0.0  0.042017  0.010504  0.033613  0.0  0.0  0.0   \n",
       "2       ...  0.061785  0.0  0.066362  0.022883  0.025172  0.0  0.0  0.0   \n",
       "3       ...  0.077720  0.0  0.049223  0.025907  0.033679  0.0  0.0  0.0   \n",
       "4       ...  0.067568  0.0  0.060811  0.013514  0.067568  0.0  0.0  0.0   \n",
       "...     ...       ...  ...       ...       ...       ...  ...  ...  ...   \n",
       "117508  ...  0.042763  0.0  0.055921  0.023026  0.075658  0.0  0.0  0.0   \n",
       "117509  ...  0.042945  0.0  0.085890  0.006135  0.015337  0.0  0.0  0.0   \n",
       "117510  ...  0.028000  0.0  0.088000  0.004000  0.032000  0.0  0.0  0.0   \n",
       "117511  ...  0.054795  0.0  0.006849  0.013699  0.041096  0.0  0.0  0.0   \n",
       "117512  ...  0.056911  0.0  0.097561  0.008130  0.056911  0.0  0.0  0.0   \n",
       "\n",
       "           classification  chainCount  \n",
       "0       TRANSPORT PROTEIN           1  \n",
       "1               HYDROLASE           4  \n",
       "2          OXIDOREDUCTASE           8  \n",
       "3               HYDROLASE           2  \n",
       "4         BINDING PROTEIN           1  \n",
       "...                   ...         ...  \n",
       "117508        TRANSFERASE           1  \n",
       "117509        TRANSFERASE           2  \n",
       "117510     OXIDOREDUCTASE           2  \n",
       "117511  SIGNALING PROTEIN           1  \n",
       "117512     OXIDOREDUCTASE           1  \n",
       "\n",
       "[117513 rows x 32 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_dict = dict(zip(list(proteins_cleaned[\"classification\"].value_counts().index), range(len(proteins_cleaned[\"classification\"].value_counts().index.tolist()))))\n",
    "proteins_cleaned[\"proteins_encoded\"] = proteins_cleaned[\"classification\"].map(proteins_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_cleaned[\"beg_seq\"] = proteins_cleaned[\"sequence\"].apply(lambda x: x[:10])\n",
    "proteins_cleaned[\"end_seq\"] = proteins_cleaned[\"sequence\"].apply(lambda x: x[-10:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "value_dict = dict(zip(proteins_cleaned[\"beg_seq\"].value_counts().keys().tolist(),\n",
    "                      range(len(proteins_cleaned[\"beg_seq\"].value_counts().index.tolist()))\n",
    "                      ))\n",
    "\n",
    "\n",
    "proteins_cleaned[\"start_seq_enc\"] = proteins_cleaned[\"beg_seq\"].map(value_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "value_dict = dict(zip(proteins_cleaned[\"end_seq\"].value_counts().keys().tolist(),\n",
    "                      range(len(proteins_cleaned[\"end_seq\"].value_counts().index.tolist()))\n",
    "                      ))\n",
    "\n",
    "\n",
    "proteins_cleaned[\"end_seq_enc\"] = proteins_cleaned[\"end_seq\"].map(value_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The RF was first used to make predictions only using the motifs as X variable and later with the entire combined dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "df_size = len(proteins_cleaned)\n",
    "train_size = int(len(proteins_cleaned)*0.7)\n",
    "\n",
    "df = proteins_cleaned.sample(frac=1)\n",
    "\n",
    "train = df[:train_size]\n",
    "test = df[train_size:]\n",
    "\n",
    "X_train = train[[\"start_seq_enc\",\"end_seq_enc\"]]\n",
    "y_train = train[\"proteins_encoded\"]\n",
    "\n",
    "X_test = test[[\"start_seq_enc\",\"end_seq_enc\"]]\n",
    "y_test = test[\"proteins_encoded\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-4 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-4 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-4 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-4 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-4 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-4 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-4 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-4 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-4 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-4 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-4 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-4 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-4 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-4 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-4 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-4\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-4\" type=\"checkbox\" checked><label for=\"sk-estimator-id-4\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier()</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = RandomForestClassifier()\n",
    "model.fit(X_train,y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = model.predict(X_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results are quite promising, they indicate that there are some patterns in the data that enable to predict protein function based on the beginning and end motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn.metrics import precision_score\n",
    "from sklearn.metrics import recall_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy: 0.6643785102399727\n",
      "F1: 0.5540940549530348\n",
      "precision: 0.5744424433817411\n",
      "recall: 0.5373828015869262\n"
     ]
    }
   ],
   "source": [
    "print(f\"accuracy: {accuracy_score(y_test, y_pred)}\")\n",
    "print(f\"F1: {f1_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"precision: {precision_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"recall: {recall_score(y_test, y_pred, average='macro')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_train = train.drop(columns=['structureId','chainId','sequence','classification','proteins_encoded', 'beg_seq', 'end_seq'])\n",
    "y_train = train[\"proteins_encoded\"]\n",
    "\n",
    "\n",
    "X_test = test.drop(columns=['structureId','chainId','sequence','classification','proteins_encoded', 'beg_seq', 'end_seq'])\n",
    "y_test = test[\"proteins_encoded\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-5 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-5 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-5 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-5 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-5 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-5 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-5 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-5 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-5 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-5 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-5 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-5 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-5 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-5 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-5 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-5 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-5 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-5 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-5\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-5\" type=\"checkbox\" checked><label for=\"sk-estimator-id-5\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier()</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = RandomForestClassifier()\n",
    "model.fit(X_train,y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The result haven't improved comparing to base model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy: 0.7648210132183582\n",
      "F1: 0.6849236923109903\n",
      "precision: 0.7943563095250471\n",
      "recall: 0.6182355059336834\n"
     ]
    }
   ],
   "source": [
    "y_pred = model.predict(X_test)\n",
    "\n",
    "print(f\"accuracy: {accuracy_score(y_test, y_pred)}\")\n",
    "print(f\"F1: {f1_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"precision: {precision_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"recall: {recall_score(y_test, y_pred, average='macro')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs = df[[\"beg_seq\",\"classification\"]].groupby(\"classification\").agg(list).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs[\"len\"]=motifs[\"beg_seq\"].apply(len)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequences do not seem to be characteristic to the group, based on quick visual comparison. There are no typical patterns that characterises particular groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>classification</th>\n",
       "      <th>beg_seq</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APOPTOSIS</td>\n",
       "      <td>[MDEDVLPGEV, AHMAHAGRSG, MGNAQERPSE, MSQSNRELV...</td>\n",
       "      <td>554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BINDING PROTEIN</td>\n",
       "      <td>[HHHHHHDMLN, GPLGSNNVND, MHHHHHHSSG, MHHHHHHSS...</td>\n",
       "      <td>9744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIOSYNTHETIC PROTEIN</td>\n",
       "      <td>[MEGTGVVAVY, MGSDKIHHHH, PPGPPGPPGP, MSNSFCVVY...</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CELL ADHESION</td>\n",
       "      <td>[KLGFFKRQYK, GSHMIWEQHT, AKPCTVSTTN, MPKPINVRV...</td>\n",
       "      <td>1386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CELL CYCLE</td>\n",
       "      <td>[MGEGFNWEIE, MASQPNSSAK, MDKPKRKEAV, ASQPNSSAK...</td>\n",
       "      <td>811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CHAPERONE</td>\n",
       "      <td>[LLKANKDLIS, MDQPMEEEEV, PMISEEREPL, VETFAFQAE...</td>\n",
       "      <td>1308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CONTRACTILE PROTEIN</td>\n",
       "      <td>[MCDEDETTAL, MDDIYKAAVE, MNPIHDRTSD, GSHMPLLSI...</td>\n",
       "      <td>292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CYTOKINE</td>\n",
       "      <td>[GAMDTHRLTR, MRKNRPAGAI, MRGSHHHHHH, SAKELRCQC...</td>\n",
       "      <td>386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FLUORESCENT PROTEIN</td>\n",
       "      <td>[GVIKPDMKIK, LYFQGMVSKG, SKGEELFTGV, MGSSLYNSH...</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HORMONE</td>\n",
       "      <td>[GIVEQCCTSI, GIVEQCCTSI, GIVEQCCASV, GSPGISGGG...</td>\n",
       "      <td>809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HYDROLASE</td>\n",
       "      <td>[TFGSGEADCG, IIGGEFTTIE, ETDEEPEEPG, MSKSVFVGE...</td>\n",
       "      <td>25783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>IMMUNE SYSTEM</td>\n",
       "      <td>[HAFQGPTSFH, EVQLVESGGG, APVRSLNCTL, DIVADHVAS...</td>\n",
       "      <td>4308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ISOMERASE</td>\n",
       "      <td>[MAALTRDPQF, GPGSMTDYIR, MSIKIRDFGL, MKVGQDKVV...</td>\n",
       "      <td>2898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LIGASE</td>\n",
       "      <td>[MGSSHHHHHH, MSLKKEYLQC, MADLSSRVNE, MAHHHHHHV...</td>\n",
       "      <td>2453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>LYASE</td>\n",
       "      <td>[TENILRKSDE, MSVRIVDVRE, MRDLKGIFSA, HWGYGKHNG...</td>\n",
       "      <td>4800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MEMBRANE PROTEIN</td>\n",
       "      <td>[GSAMGSSDEP, GSEKWTSTAI, MASEFKKKLF, GSQGERETW...</td>\n",
       "      <td>1970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>MOTOR PROTEIN</td>\n",
       "      <td>[MENLISLVNK, KQQEVIADKQ, GSKGNIRVIA, MRECISIHV...</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>[AETTVKPTKL, MAASCVLLHT, MAHHHHHHKK, ANPCCSNPC...</td>\n",
       "      <td>13575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>OXYGEN STORAGE</td>\n",
       "      <td>[EQWQTLYEAI, PSVYDAAAQL, MVLSEGEWQL, VLSPADKTN...</td>\n",
       "      <td>462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>PHOTOSYNTHESIS</td>\n",
       "      <td>[MEHVAFGSED, MTTTLQRRES, MKTPITEAIA, ANLWERFCN...</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>PROTEIN FIBRIL</td>\n",
       "      <td>[SSTNVG, SSTSAA, NNQQNY, AILSST, VQIVYKPVDL, D...</td>\n",
       "      <td>151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RECEPTOR</td>\n",
       "      <td>[GAPVPVDEND, MGSSHHHHHH, GALNPESADL, EPKIHYRKD...</td>\n",
       "      <td>768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>REGULATORY PROTEIN</td>\n",
       "      <td>[MHHHHHHSSG, MHHHHHHSSG, MTKRIASVEK, GPLVPESLQ...</td>\n",
       "      <td>1431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SIGNALING PROTEIN</td>\n",
       "      <td>[KFSPQLLSLL, HMGSMALDGI, GSSGSSGMAL, GPLGSPEFP...</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>STRUCTURAL GENOMICS</td>\n",
       "      <td>[VRIRRAGLED, MSLNTPVSVN, GDEQSEAQFF, MTYEAFVEL...</td>\n",
       "      <td>2592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>STRUCTURAL PROTEIN</td>\n",
       "      <td>[LQIFVKTLTG, PTKLETFPLD, SHIDCRMSPW, PHMQPFDSG...</td>\n",
       "      <td>1353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TOXIN</td>\n",
       "      <td>[VHKMAKNQFG, MPVNIKNFNY, PVTINNFNYN, IEAIRCGGS...</td>\n",
       "      <td>1250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TRANSCRIPTION</td>\n",
       "      <td>[GSMNKHTEHD, HHHHHHMNSN, MATVSVSLKS, IGIPFPDHS...</td>\n",
       "      <td>3869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TRANSFERASE</td>\n",
       "      <td>[MHHHHHHVMI, MNDIVASTQL, IRTEKIICRD, MSQPSTANG...</td>\n",
       "      <td>18884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TRANSLATION</td>\n",
       "      <td>[MPEICPRCGL, MAWPKVQPEV, KPYPCGLCNR, MHHHHHHSS...</td>\n",
       "      <td>462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>TRANSPORT PROTEIN</td>\n",
       "      <td>[SSQIRQNYST, VLSAADKNNV, ADLALGAQVF, MNHKVHMDG...</td>\n",
       "      <td>7167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>VIRAL PROTEIN</td>\n",
       "      <td>[AITIGNKNDD, ADPGATLCLG, GFNLKVKDLN, ADPGATLCL...</td>\n",
       "      <td>2719</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>VIRUS</td>\n",
       "      <td>[TFFKTVYRRY, MAKLETVTLG, TTSAGESADP, VGNSSGNWH...</td>\n",
       "      <td>587</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          classification                                            beg_seq  \\\n",
       "0              APOPTOSIS  [MDEDVLPGEV, AHMAHAGRSG, MGNAQERPSE, MSQSNRELV...   \n",
       "1        BINDING PROTEIN  [HHHHHHDMLN, GPLGSNNVND, MHHHHHHSSG, MHHHHHHSS...   \n",
       "2   BIOSYNTHETIC PROTEIN  [MEGTGVVAVY, MGSDKIHHHH, PPGPPGPPGP, MSNSFCVVY...   \n",
       "3          CELL ADHESION  [KLGFFKRQYK, GSHMIWEQHT, AKPCTVSTTN, MPKPINVRV...   \n",
       "4             CELL CYCLE  [MGEGFNWEIE, MASQPNSSAK, MDKPKRKEAV, ASQPNSSAK...   \n",
       "5              CHAPERONE  [LLKANKDLIS, MDQPMEEEEV, PMISEEREPL, VETFAFQAE...   \n",
       "6    CONTRACTILE PROTEIN  [MCDEDETTAL, MDDIYKAAVE, MNPIHDRTSD, GSHMPLLSI...   \n",
       "7               CYTOKINE  [GAMDTHRLTR, MRKNRPAGAI, MRGSHHHHHH, SAKELRCQC...   \n",
       "8    FLUORESCENT PROTEIN  [GVIKPDMKIK, LYFQGMVSKG, SKGEELFTGV, MGSSLYNSH...   \n",
       "9                HORMONE  [GIVEQCCTSI, GIVEQCCTSI, GIVEQCCASV, GSPGISGGG...   \n",
       "10             HYDROLASE  [TFGSGEADCG, IIGGEFTTIE, ETDEEPEEPG, MSKSVFVGE...   \n",
       "11         IMMUNE SYSTEM  [HAFQGPTSFH, EVQLVESGGG, APVRSLNCTL, DIVADHVAS...   \n",
       "12             ISOMERASE  [MAALTRDPQF, GPGSMTDYIR, MSIKIRDFGL, MKVGQDKVV...   \n",
       "13                LIGASE  [MGSSHHHHHH, MSLKKEYLQC, MADLSSRVNE, MAHHHHHHV...   \n",
       "14                 LYASE  [TENILRKSDE, MSVRIVDVRE, MRDLKGIFSA, HWGYGKHNG...   \n",
       "15      MEMBRANE PROTEIN  [GSAMGSSDEP, GSEKWTSTAI, MASEFKKKLF, GSQGERETW...   \n",
       "16         MOTOR PROTEIN  [MENLISLVNK, KQQEVIADKQ, GSKGNIRVIA, MRECISIHV...   \n",
       "17        OXIDOREDUCTASE  [AETTVKPTKL, MAASCVLLHT, MAHHHHHHKK, ANPCCSNPC...   \n",
       "18        OXYGEN STORAGE  [EQWQTLYEAI, PSVYDAAAQL, MVLSEGEWQL, VLSPADKTN...   \n",
       "19        PHOTOSYNTHESIS  [MEHVAFGSED, MTTTLQRRES, MKTPITEAIA, ANLWERFCN...   \n",
       "20        PROTEIN FIBRIL  [SSTNVG, SSTSAA, NNQQNY, AILSST, VQIVYKPVDL, D...   \n",
       "21              RECEPTOR  [GAPVPVDEND, MGSSHHHHHH, GALNPESADL, EPKIHYRKD...   \n",
       "22    REGULATORY PROTEIN  [MHHHHHHSSG, MHHHHHHSSG, MTKRIASVEK, GPLVPESLQ...   \n",
       "23     SIGNALING PROTEIN  [KFSPQLLSLL, HMGSMALDGI, GSSGSSGMAL, GPLGSPEFP...   \n",
       "24   STRUCTURAL GENOMICS  [VRIRRAGLED, MSLNTPVSVN, GDEQSEAQFF, MTYEAFVEL...   \n",
       "25    STRUCTURAL PROTEIN  [LQIFVKTLTG, PTKLETFPLD, SHIDCRMSPW, PHMQPFDSG...   \n",
       "26                 TOXIN  [VHKMAKNQFG, MPVNIKNFNY, PVTINNFNYN, IEAIRCGGS...   \n",
       "27         TRANSCRIPTION  [GSMNKHTEHD, HHHHHHMNSN, MATVSVSLKS, IGIPFPDHS...   \n",
       "28           TRANSFERASE  [MHHHHHHVMI, MNDIVASTQL, IRTEKIICRD, MSQPSTANG...   \n",
       "29           TRANSLATION  [MPEICPRCGL, MAWPKVQPEV, KPYPCGLCNR, MHHHHHHSS...   \n",
       "30     TRANSPORT PROTEIN  [SSQIRQNYST, VLSAADKNNV, ADLALGAQVF, MNHKVHMDG...   \n",
       "31         VIRAL PROTEIN  [AITIGNKNDD, ADPGATLCLG, GFNLKVKDLN, ADPGATLCL...   \n",
       "32                 VIRUS  [TFFKTVYRRY, MAKLETVTLG, TTSAGESADP, VGNSSGNWH...   \n",
       "\n",
       "      len  \n",
       "0     554  \n",
       "1    9744  \n",
       "2     438  \n",
       "3    1386  \n",
       "4     811  \n",
       "5    1308  \n",
       "6     292  \n",
       "7     386  \n",
       "8     369  \n",
       "9     809  \n",
       "10  25783  \n",
       "11   4308  \n",
       "12   2898  \n",
       "13   2453  \n",
       "14   4800  \n",
       "15   1970  \n",
       "16    307  \n",
       "17  13575  \n",
       "18    462  \n",
       "19    275  \n",
       "20    151  \n",
       "21    768  \n",
       "22   1431  \n",
       "23   3352  \n",
       "24   2592  \n",
       "25   1353  \n",
       "26   1250  \n",
       "27   3869  \n",
       "28  18884  \n",
       "29    462  \n",
       "30   7167  \n",
       "31   2719  \n",
       "32    587  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another approach was to cluster similar sequences together. To simplify the process and make it more computationally efficient only 5 amino acids were sliced both from the beginning and the end of the sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_cleaned[\"beg_seq\"] = proteins_cleaned[\"beg_seq\"].apply(lambda x: x[:5])\n",
    "proteins_cleaned[\"end_seq\"] = proteins_cleaned[\"end_seq\"].apply(lambda x: x[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>structureId</th>\n",
       "      <th>chainId</th>\n",
       "      <th>sequence</th>\n",
       "      <th>residueCount</th>\n",
       "      <th>len</th>\n",
       "      <th>A</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>...</th>\n",
       "      <th>X</th>\n",
       "      <th>B</th>\n",
       "      <th>Z</th>\n",
       "      <th>classification</th>\n",
       "      <th>chainCount</th>\n",
       "      <th>proteins_encoded</th>\n",
       "      <th>beg_seq</th>\n",
       "      <th>end_seq</th>\n",
       "      <th>start_seq_enc</th>\n",
       "      <th>end_seq_enc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4LZV</td>\n",
       "      <td>A</td>\n",
       "      <td>LIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEEL...</td>\n",
       "      <td>162</td>\n",
       "      <td>162</td>\n",
       "      <td>0.092593</td>\n",
       "      <td>0.030864</td>\n",
       "      <td>0.061728</td>\n",
       "      <td>0.098765</td>\n",
       "      <td>0.024691</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSPORT PROTEIN</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>LIVTQ</td>\n",
       "      <td>PTQLE</td>\n",
       "      <td>177</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4FK5</td>\n",
       "      <td>A</td>\n",
       "      <td>GAAAAMSICPHIQQVFQNEKSKDGVLKTCNAARYILNHSVPKEKFL...</td>\n",
       "      <td>767</td>\n",
       "      <td>476</td>\n",
       "      <td>0.042017</td>\n",
       "      <td>0.054622</td>\n",
       "      <td>0.048319</td>\n",
       "      <td>0.048319</td>\n",
       "      <td>0.052521</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HYDROLASE</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>GAAAA</td>\n",
       "      <td>LLFYT</td>\n",
       "      <td>2113</td>\n",
       "      <td>2045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4DOY</td>\n",
       "      <td>A</td>\n",
       "      <td>MGSSHHHHHHSSGLVPRGSHMTLSPEKQHVRPRDAADNDPVAVARG...</td>\n",
       "      <td>3496</td>\n",
       "      <td>437</td>\n",
       "      <td>0.125858</td>\n",
       "      <td>0.002288</td>\n",
       "      <td>0.059497</td>\n",
       "      <td>0.054920</td>\n",
       "      <td>0.032037</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>MGSSH</td>\n",
       "      <td>GQYPI</td>\n",
       "      <td>0</td>\n",
       "      <td>4462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2XZK</td>\n",
       "      <td>A</td>\n",
       "      <td>INDPAKSAAPYHDEFPLFRSANMASPDKLSTGIGFHSFRIPAVVRT...</td>\n",
       "      <td>772</td>\n",
       "      <td>386</td>\n",
       "      <td>0.085492</td>\n",
       "      <td>0.002591</td>\n",
       "      <td>0.072539</td>\n",
       "      <td>0.033679</td>\n",
       "      <td>0.036269</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>HYDROLASE</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>INDPA</td>\n",
       "      <td>LSWIL</td>\n",
       "      <td>4064</td>\n",
       "      <td>2825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4WR9</td>\n",
       "      <td>A</td>\n",
       "      <td>AYLDEELQTELYEIKHQILQTMGVLSLQGSMLSVGDKVFSTNGQSV...</td>\n",
       "      <td>148</td>\n",
       "      <td>148</td>\n",
       "      <td>0.054054</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.087838</td>\n",
       "      <td>0.027027</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BINDING PROTEIN</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>AYLDE</td>\n",
       "      <td>LQYRL</td>\n",
       "      <td>736</td>\n",
       "      <td>986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117508</th>\n",
       "      <td>3QAH</td>\n",
       "      <td>A</td>\n",
       "      <td>MGSHHHHHHHHGSDYDIPTTENLYFQGSTKVKYVDKIHIGNYEIDA...</td>\n",
       "      <td>304</td>\n",
       "      <td>304</td>\n",
       "      <td>0.023026</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>0.052632</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.039474</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSFERASE</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>MGSHH</td>\n",
       "      <td>SVCLK</td>\n",
       "      <td>753</td>\n",
       "      <td>4586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117509</th>\n",
       "      <td>4M0P</td>\n",
       "      <td>A</td>\n",
       "      <td>MPNIKIFSGSSHQDLSQKIADRLGLELGKVVTKKFSNQETCVEIGE...</td>\n",
       "      <td>652</td>\n",
       "      <td>326</td>\n",
       "      <td>0.095092</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>0.067485</td>\n",
       "      <td>0.052147</td>\n",
       "      <td>0.027607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TRANSFERASE</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>MPNIK</td>\n",
       "      <td>PLLEH</td>\n",
       "      <td>1026</td>\n",
       "      <td>387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117510</th>\n",
       "      <td>4NPM</td>\n",
       "      <td>A</td>\n",
       "      <td>DESEYEERRDAEARRVKSGIKQASIFTLEECARIEAKIDEVVAKAD...</td>\n",
       "      <td>500</td>\n",
       "      <td>250</td>\n",
       "      <td>0.072000</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.068000</td>\n",
       "      <td>0.076000</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>DESEY</td>\n",
       "      <td>LKAKR</td>\n",
       "      <td>14903</td>\n",
       "      <td>13361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117511</th>\n",
       "      <td>2AF0</td>\n",
       "      <td>A</td>\n",
       "      <td>ADLGTENLYFQSMKPSPEEAQLWSEAFDELLASKYGLAAFRAFLKS...</td>\n",
       "      <td>146</td>\n",
       "      <td>146</td>\n",
       "      <td>0.089041</td>\n",
       "      <td>0.027397</td>\n",
       "      <td>0.041096</td>\n",
       "      <td>0.116438</td>\n",
       "      <td>0.082192</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>SIGNALING PROTEIN</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>ADLGT</td>\n",
       "      <td>FYQDL</td>\n",
       "      <td>15693</td>\n",
       "      <td>23262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117512</th>\n",
       "      <td>2XDP</td>\n",
       "      <td>A</td>\n",
       "      <td>SMCEKVISVGQTVITKHRNTRYYSCRVMAVTSQTFYEVMFDDGSFS...</td>\n",
       "      <td>123</td>\n",
       "      <td>123</td>\n",
       "      <td>0.040650</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.073171</td>\n",
       "      <td>0.081301</td>\n",
       "      <td>0.048780</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OXIDOREDUCTASE</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>SMCEK</td>\n",
       "      <td>LDEEL</td>\n",
       "      <td>41470</td>\n",
       "      <td>39788</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117513 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       structureId chainId                                           sequence  \\\n",
       "0             4LZV       A  LIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEEL...   \n",
       "1             4FK5       A  GAAAAMSICPHIQQVFQNEKSKDGVLKTCNAARYILNHSVPKEKFL...   \n",
       "2             4DOY       A  MGSSHHHHHHSSGLVPRGSHMTLSPEKQHVRPRDAADNDPVAVARG...   \n",
       "3             2XZK       A  INDPAKSAAPYHDEFPLFRSANMASPDKLSTGIGFHSFRIPAVVRT...   \n",
       "4             4WR9       A  AYLDEELQTELYEIKHQILQTMGVLSLQGSMLSVGDKVFSTNGQSV...   \n",
       "...            ...     ...                                                ...   \n",
       "117508        3QAH       A  MGSHHHHHHHHGSDYDIPTTENLYFQGSTKVKYVDKIHIGNYEIDA...   \n",
       "117509        4M0P       A  MPNIKIFSGSSHQDLSQKIADRLGLELGKVVTKKFSNQETCVEIGE...   \n",
       "117510        4NPM       A  DESEYEERRDAEARRVKSGIKQASIFTLEECARIEAKIDEVVAKAD...   \n",
       "117511        2AF0       A  ADLGTENLYFQSMKPSPEEAQLWSEAFDELLASKYGLAAFRAFLKS...   \n",
       "117512        2XDP       A  SMCEKVISVGQTVITKHRNTRYYSCRVMAVTSQTFYEVMFDDGSFS...   \n",
       "\n",
       "        residueCount  len         A         C         D         E         F  \\\n",
       "0                162  162  0.092593  0.030864  0.061728  0.098765  0.024691   \n",
       "1                767  476  0.042017  0.054622  0.048319  0.048319  0.052521   \n",
       "2               3496  437  0.125858  0.002288  0.059497  0.054920  0.032037   \n",
       "3                772  386  0.085492  0.002591  0.072539  0.033679  0.036269   \n",
       "4                148  148  0.054054  0.027027  0.067568  0.087838  0.027027   \n",
       "...              ...  ...       ...       ...       ...       ...       ...   \n",
       "117508           304  304  0.023026  0.026316  0.052632  0.055921  0.039474   \n",
       "117509           652  326  0.095092  0.027607  0.067485  0.052147  0.027607   \n",
       "117510           500  250  0.072000  0.020000  0.068000  0.076000  0.040000   \n",
       "117511           146  146  0.089041  0.027397  0.041096  0.116438  0.082192   \n",
       "117512           123  123  0.040650  0.024390  0.073171  0.081301  0.048780   \n",
       "\n",
       "        ...    X    B    Z     classification  chainCount  proteins_encoded  \\\n",
       "0       ...  0.0  0.0  0.0  TRANSPORT PROTEIN           1                 4   \n",
       "1       ...  0.0  0.0  0.0          HYDROLASE           4                 0   \n",
       "2       ...  0.0  0.0  0.0     OXIDOREDUCTASE           8                 2   \n",
       "3       ...  0.0  0.0  0.0          HYDROLASE           2                 0   \n",
       "4       ...  0.0  0.0  0.0    BINDING PROTEIN           1                 3   \n",
       "...     ...  ...  ...  ...                ...         ...               ...   \n",
       "117508  ...  0.0  0.0  0.0        TRANSFERASE           1                 1   \n",
       "117509  ...  0.0  0.0  0.0        TRANSFERASE           2                 1   \n",
       "117510  ...  0.0  0.0  0.0     OXIDOREDUCTASE           2                 2   \n",
       "117511  ...  0.0  0.0  0.0  SIGNALING PROTEIN           1                 8   \n",
       "117512  ...  0.0  0.0  0.0     OXIDOREDUCTASE           1                 2   \n",
       "\n",
       "        beg_seq  end_seq  start_seq_enc  end_seq_enc  \n",
       "0         LIVTQ    PTQLE            177          165  \n",
       "1         GAAAA    LLFYT           2113         2045  \n",
       "2         MGSSH    GQYPI              0         4462  \n",
       "3         INDPA    LSWIL           4064         2825  \n",
       "4         AYLDE    LQYRL            736          986  \n",
       "...         ...      ...            ...          ...  \n",
       "117508    MGSHH    SVCLK            753         4586  \n",
       "117509    MPNIK    PLLEH           1026          387  \n",
       "117510    DESEY    LKAKR          14903        13361  \n",
       "117511    ADLGT    FYQDL          15693        23262  \n",
       "117512    SMCEK    LDEEL          41470        39788  \n",
       "\n",
       "[117513 rows x 37 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "proteins_cleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_beg = pd.DataFrame({\"beg_seq\": proteins_cleaned['beg_seq'].value_counts().index, \"count\":proteins_cleaned[\"beg_seq\"].value_counts().values})\n",
    "sequences_end = pd.DataFrame({\"end_seq\": proteins_cleaned['end_seq'].value_counts().index, \"count\":proteins_cleaned[\"end_seq\"].value_counts().values})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_filtered_beg = sequences_beg[sequences_beg[\"count\"]>=100]\n",
    "sequences_filtered_end = sequences_end[sequences_end[\"count\"]>=100]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next step was to identify repetitive sequences within the dataset. From the sliced parts the occurrences of identical motifs were counted. It unfortunately resulted in relatively low values, with many unique sequences and numerous instances repeating only a few hundred times, which is not significant in a dataset of 100 000 instances. Despite this, the approach was tested to determine whether sequences could be grouped into similar motifs that might still be characteristic of specific proteins.  \n",
    "\n",
    "Repetitive motifs were found in only around 1/3 of the instances in the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(4703)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_filtered_beg[\"count\"].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(27658)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_filtered_beg[\"count\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(12541)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences_filtered_end[\"count\"].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "beg_motifs_df = pd.DataFrame(columns=np.array(sequences_filtered_beg[\"beg_seq\"]), index=range(len(sequences_filtered_beg)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Similarity between sequences was calculated using local sequence alignment from Biopython package. Then similarity matrix was constructed and calculated pairwise similarity. Finally a threshold (0,6) was established to define if sequences are similar to each other. \n",
    "\n",
    "On the resulting matrix connected component algorithm was executed to identify groups of similar motifs. They were labeled as one group. Motifs at the beginning and end of the sequences were identified and assigned corresponding labels. While this approach may not be ideal, as it could imply a numerical order, it was the simplest solution at this stage. Therefore, this encoding was used for the initial test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def jaccard_similarity(target,query):\n",
    "#     set_a = set(target)\n",
    "#     set_b = set(query)\n",
    "#     score = len(set_a & set_b) / len(set_a| set_b)\n",
    "#     return score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "aligner = Align.PairwiseAligner()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx_col, column in enumerate(beg_motifs_df.columns):\n",
    "\n",
    "    for idx_row, row in beg_motifs_df.iterrows():\n",
    "        target = beg_motifs_df.columns[idx_col]\n",
    "        query = beg_motifs_df.columns[idx_row]\n",
    "        score = aligner.score(target, query) / (max(len(target),len(query)))\n",
    "\n",
    "        beg_motifs_df.iloc[idx_row,idx_col] = score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGSSH</th>\n",
       "      <th>MHHHH</th>\n",
       "      <th>GSSGS</th>\n",
       "      <th>MAHHH</th>\n",
       "      <th>GPLGS</th>\n",
       "      <th>HHHHH</th>\n",
       "      <th>MRGSH</th>\n",
       "      <th>MGHHH</th>\n",
       "      <th>KVFGR</th>\n",
       "      <th>PQITL</th>\n",
       "      <th>...</th>\n",
       "      <th>GSHMK</th>\n",
       "      <th>GAMDP</th>\n",
       "      <th>ENSNI</th>\n",
       "      <th>VLSEG</th>\n",
       "      <th>GANKT</th>\n",
       "      <th>DYKDD</th>\n",
       "      <th>SNAMN</th>\n",
       "      <th>MGSHH</th>\n",
       "      <th>IVGGQ</th>\n",
       "      <th>GLSDG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>0.8</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.8</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>0.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.6</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.2</td>\n",
       "      <td>0.4</td>\n",
       "      <td>0.4</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   MGSSH MHHHH GSSGS MAHHH GPLGS HHHHH MRGSH MGHHH KVFGR PQITL  ... GSHMK  \\\n",
       "0    1.0   0.4   0.6   0.4   0.4   0.2   0.8   0.6   0.2   0.0  ...   0.6   \n",
       "1    0.4   1.0   0.0   0.8   0.0   0.8   0.4   0.8   0.0   0.0  ...   0.2   \n",
       "2    0.6   0.0   1.0   0.0   0.6   0.0   0.4   0.2   0.2   0.0  ...   0.4   \n",
       "3    0.4   0.8   0.0   1.0   0.0   0.6   0.4   0.8   0.0   0.0  ...   0.2   \n",
       "4    0.4   0.0   0.6   0.0   1.0   0.0   0.4   0.2   0.2   0.4  ...   0.4   \n",
       "..   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   \n",
       "63   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.2   0.0  ...   0.2   \n",
       "64   0.2   0.2   0.2   0.2   0.2   0.0   0.2   0.2   0.0   0.0  ...   0.4   \n",
       "65   0.8   0.6   0.4   0.6   0.4   0.4   0.8   0.8   0.2   0.0  ...   0.6   \n",
       "66   0.2   0.0   0.4   0.0   0.4   0.0   0.2   0.2   0.4   0.2  ...   0.2   \n",
       "67   0.4   0.0   0.6   0.0   0.6   0.0   0.4   0.2   0.2   0.2  ...   0.4   \n",
       "\n",
       "   GAMDP ENSNI VLSEG GANKT DYKDD SNAMN MGSHH IVGGQ GLSDG  \n",
       "0    0.2   0.2   0.2   0.2   0.0   0.2   0.8   0.2   0.4  \n",
       "1    0.2   0.0   0.0   0.0   0.0   0.2   0.6   0.0   0.0  \n",
       "2    0.2   0.2   0.4   0.2   0.0   0.2   0.4   0.4   0.6  \n",
       "3    0.2   0.0   0.0   0.2   0.0   0.2   0.6   0.0   0.0  \n",
       "4    0.4   0.2   0.4   0.2   0.0   0.2   0.4   0.4   0.6  \n",
       "..   ...   ...   ...   ...   ...   ...   ...   ...   ...  \n",
       "63   0.2   0.0   0.0   0.2   1.0   0.0   0.0   0.0   0.2  \n",
       "64   0.4   0.4   0.2   0.4   0.0   1.0   0.2   0.0   0.2  \n",
       "65   0.2   0.2   0.2   0.2   0.0   0.2   1.0   0.2   0.4  \n",
       "66   0.2   0.2   0.4   0.2   0.0   0.0   0.2   1.0   0.4  \n",
       "67   0.4   0.2   0.6   0.2   0.2   0.2   0.4   0.4   1.0  \n",
       "\n",
       "[68 rows x 68 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beg_motifs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "beg_motifs_df_binary = beg_motifs_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "for column in beg_motifs_df_binary.columns:\n",
    "    beg_motifs_df_binary[column] = beg_motifs_df_binary[column].apply(lambda x: 1 if x>=0.6 else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from scipy.sparse.csgraph import connected_components\n",
    "\n",
    "n_components, labels = connected_components(csgraph=np.array(beg_motifs_df_binary), directed=False, return_labels=True)\n",
    "n_components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "beg_motifs_df_binary.index = beg_motifs_df_binary.columns\n",
    "beg_motifs_df_binary[\"labels\"] = labels\n",
    "beg_motifs_df_binary[\"seq\"] = beg_motifs_df_binary.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGSSH</th>\n",
       "      <th>MHHHH</th>\n",
       "      <th>GSSGS</th>\n",
       "      <th>MAHHH</th>\n",
       "      <th>GPLGS</th>\n",
       "      <th>HHHHH</th>\n",
       "      <th>MRGSH</th>\n",
       "      <th>MGHHH</th>\n",
       "      <th>KVFGR</th>\n",
       "      <th>PQITL</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSNI</th>\n",
       "      <th>VLSEG</th>\n",
       "      <th>GANKT</th>\n",
       "      <th>DYKDD</th>\n",
       "      <th>SNAMN</th>\n",
       "      <th>MGSHH</th>\n",
       "      <th>IVGGQ</th>\n",
       "      <th>GLSDG</th>\n",
       "      <th>labels</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MGSSH</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MGSSH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MHHHH</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MHHHH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSSGS</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>GSSGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAHHH</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MAHHH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GPLGS</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>GPLGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYKDD</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>DYKDD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNAMN</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>SNAMN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGSHH</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>MGSHH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IVGGQ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>IVGGQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GLSDG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>GLSDG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>68 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       MGSSH  MHHHH  GSSGS  MAHHH  GPLGS  HHHHH  MRGSH  MGHHH  KVFGR  PQITL  \\\n",
       "MGSSH      1      0      1      0      0      0      1      1      0      0   \n",
       "MHHHH      0      1      0      1      0      1      0      1      0      0   \n",
       "GSSGS      1      0      1      0      1      0      0      0      0      0   \n",
       "MAHHH      0      1      0      1      0      1      0      1      0      0   \n",
       "GPLGS      0      0      1      0      1      0      0      0      0      0   \n",
       "...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "DYKDD      0      0      0      0      0      0      0      0      0      0   \n",
       "SNAMN      0      0      0      0      0      0      0      0      0      0   \n",
       "MGSHH      1      1      0      1      0      0      1      1      0      0   \n",
       "IVGGQ      0      0      0      0      0      0      0      0      0      0   \n",
       "GLSDG      0      0      1      0      1      0      0      0      0      0   \n",
       "\n",
       "       ...  ENSNI  VLSEG  GANKT  DYKDD  SNAMN  MGSHH  IVGGQ  GLSDG  labels  \\\n",
       "MGSSH  ...      0      0      0      0      0      1      0      0       0   \n",
       "MHHHH  ...      0      0      0      0      0      1      0      0       0   \n",
       "GSSGS  ...      0      0      0      0      0      0      0      1       0   \n",
       "MAHHH  ...      0      0      0      0      0      1      0      0       0   \n",
       "GPLGS  ...      0      0      0      0      0      0      0      1       0   \n",
       "...    ...    ...    ...    ...    ...    ...    ...    ...    ...     ...   \n",
       "DYKDD  ...      0      0      0      1      0      0      0      0      11   \n",
       "SNAMN  ...      0      0      0      0      1      0      0      0       0   \n",
       "MGSHH  ...      0      0      0      0      0      1      0      0       0   \n",
       "IVGGQ  ...      0      0      0      0      0      0      1      0       0   \n",
       "GLSDG  ...      0      1      0      0      0      0      0      1       0   \n",
       "\n",
       "         seq  \n",
       "MGSSH  MGSSH  \n",
       "MHHHH  MHHHH  \n",
       "GSSGS  GSSGS  \n",
       "MAHHH  MAHHH  \n",
       "GPLGS  GPLGS  \n",
       "...      ...  \n",
       "DYKDD  DYKDD  \n",
       "SNAMN  SNAMN  \n",
       "MGSHH  MGSHH  \n",
       "IVGGQ  IVGGQ  \n",
       "GLSDG  GLSDG  \n",
       "\n",
       "[68 rows x 70 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "beg_motifs_df_binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "beg_seq_dict = dict(zip(beg_motifs_df_binary[\"seq\"], beg_motifs_df_binary[\"labels\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "end_motifs_df = pd.DataFrame(columns=np.array(sequences_filtered_end[\"end_seq\"]), index=range(len(sequences_filtered_end)))\n",
    "\n",
    "for idx_col, column in enumerate(end_motifs_df.columns):\n",
    "\n",
    "    for idx_row, row in end_motifs_df.iterrows():\n",
    "        target = end_motifs_df.columns[idx_col]\n",
    "        query = end_motifs_df.columns[idx_row]\n",
    "        score = aligner.score(target, query) / (max(len(target),len(query)))\n",
    "\n",
    "        end_motifs_df.iloc[idx_row,idx_col] = score\n",
    "\n",
    "end_motifs_df_binary = end_motifs_df.copy()\n",
    "\n",
    "for column in end_motifs_df_binary.columns:\n",
    "    end_motifs_df_binary[column] = end_motifs_df_binary[column].apply(lambda x: 1 if x>=0.6 else 0)\n",
    "\n",
    "\n",
    "n_components, labels = connected_components(csgraph=np.array(end_motifs_df_binary), directed=False, return_labels=True)\n",
    "\n",
    "\n",
    "end_motifs_df_binary.index = end_motifs_df_binary.columns\n",
    "end_motifs_df_binary[\"labels\"] = labels\n",
    "end_motifs_df_binary[\"seq\"] = end_motifs_df_binary.index\n",
    "\n",
    "end_seq_dict = dict(zip(end_motifs_df_binary[\"seq\"], end_motifs_df_binary[\"labels\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_cleaned[\"beg_motif\"] = proteins_cleaned[\"beg_seq\"].map(beg_seq_dict)\n",
    "proteins_cleaned[\"end_motif\"] = proteins_cleaned[\"end_seq\"].map(end_seq_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "proteins_cleaned[\"beg_motif\"].fillna(-1, inplace=True)\n",
    "proteins_cleaned[\"end_motif\"].fillna(-1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Final test using Random Forest were performed and achieved very similar results to the base model. It is possible that the few first nucleotides do not provide enough information, cases belonging to any identified group constitute only around 25% of the dataset and as a result the information is probably insufficient. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = proteins_cleaned.sample(frac=1)\n",
    "\n",
    "train = df[:train_size]\n",
    "test = df[train_size:]\n",
    "\n",
    "X_train = train.drop(columns=['structureId','chainId','sequence','classification','proteins_encoded', 'beg_seq', 'end_seq','start_seq_enc', 'end_seq_enc'])\n",
    "y_train = train[\"proteins_encoded\"]\n",
    "\n",
    "X_test = test.drop(columns=['structureId','chainId','sequence','classification','proteins_encoded', 'beg_seq', 'end_seq','start_seq_enc', 'end_seq_enc'])\n",
    "y_test = test[\"proteins_encoded\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['residueCount', 'len', 'A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L',\n",
       "       'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'Y', 'X', 'B',\n",
       "       'Z', 'chainCount', 'beg_motif', 'end_motif'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  max-height: 0;\n",
       "  max-width: 0;\n",
       "  overflow: hidden;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  max-height: 200px;\n",
       "  max-width: 100%;\n",
       "  overflow: auto;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>RandomForestClassifier</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.6/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier()</pre></div> </div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = RandomForestClassifier()\n",
    "model.fit(X_train,y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy: 0.7689907528223747\n",
      "F1: 0.6944700357592838\n",
      "precision: 0.8010852674526864\n",
      "recall: 0.6286207505610867\n"
     ]
    }
   ],
   "source": [
    "y_pred = model.predict(X_test)\n",
    "\n",
    "print(f\"accuracy: {accuracy_score(y_test, y_pred)}\")\n",
    "print(f\"F1: {f1_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"precision: {precision_score(y_test, y_pred, average='macro')}\")\n",
    "print(f\"recall: {recall_score(y_test, y_pred, average='macro')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import json\n",
    "\n",
    "# with open('./Data/beg_motifs_dict.json', 'w') as fp:\n",
    "#     json.dump(beg_seq_dict, fp)\n",
    "\n",
    "# with open('./Data/end_motifs_dict.json', 'w') as fp:\n",
    "#     json.dump(end_seq_dict, fp)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
